### PR TITLE
[test-improver] test: add edge case tests for TestContextShouldBeValidAnalyzer (MSTEST0005)

### PR DESCRIPTION
**Pull request by Evangelink**
## Goal and Rationale

`TestContextShouldBeValidAnalyzer` (MSTEST0005) has non-trivial logic across several paths — `IsTestContextPropertyAutomaticallyAssigned`, `TryGetTestContextParameterIfValidConstructor`, and the early-return for non-test-class types — that had no dedicated test coverage.

## Approach

Added 4 targeted edge-case tests covering previously untested code paths:

| New test | Code path exercised |
|---|---|
| `WhenTestContextPropertyHasPrivateSetter_NoDiagnostic` | `IsTestContextPropertyAutomaticallyAssigned` → `SetMethod is not null` with a private setter |
| `WhenTestContextPropertyHasProtectedSetter_NoDiagnostic` | Same check with a protected setter |
| `WhenInvalidTestContextPropertyIsOnNonTestClass_NoDiagnostic` | Early return: class without `[TestClass]` attribute is skipped entirely |
| `WhenTestContextAssignedOnlyInMultiParamConstructor_Diagnostic` | `TryGetTestContextParameterIfValidConstructor` only accepts single-parameter constructors; a two-parameter constructor is not recognised, so the readonly property is correctly reported |

## Trade-offs

- Tests are self-contained; no new dependencies or infrastructure.
- All 4 cases map to distinct, understandable behaviors that developers could reasonably encounter.

## Reproducibility

```bash
DOTNET_INSTALL_DIR=/usr/share/dotnet ./build.sh --restore
.dotnet/dotnet run --project test/UnitTests/MSTest.Analyzers.UnitTests -f net8.0 --no-build -c Debug -- --filter "ClassName~TestContextShouldBeValidAnalyzerTests"
```

## Test Status

✅ 71 tests passed (67 existing + 4 new), 0 failed.

> 🤖 **Automated content by GitHub Copilot.** Posted via a maintainer's GitHub token, so it appears under their account — the account owner did **not** write or approve this content personally. Generated by the [Test Improver](https://github.com/microsoft/testfx/actions/runs/27726042801/agentic_workflow) workflow. · 1.1K AIC · ⌖ 40.3 AIC · [◷]( · [◷](https://github.com/search?q=repo%3Amicrosoft%2Ftestfx+%22gh-aw-workflow-id%3A+test-improver%22&type=pullrequests))
>
<details>
<summary>Add this agentic workflows to your repo</summary>

To install this agentic workflow, run

```
gh aw add githubnext/agentics/workflows/test-improver.md@main
```
</details>

### CODE REVIEW

**Comment by Evangelink** [comment]
### 🧪 Test quality grade — PR #9223

<table>
  <thead>
    <tr><th>Δ</th><th>Test</th><th>Grade</th><th>Band</th><th>Notes</th></tr>
  </thead>
  <tbody>
    <tr>
      <td>new</td>
      <td><code>TestContextShouldBeValidAnalyzerTests.<br>WhenTestContextAssignedOnlyInMultiParamConstructor_<br>Diagnostic</code></td>
      <td>B</td>
      <td>80–89</td>
      <td>Mildly long body (37 lines) from two inline code snippets; single behavior, strong assertion on diagnostic location and code fix.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>TestContextShouldBeValidAnalyzerTests.<br>WhenInvalidTestContextPropertyIsOnNonTestClass_<br>NoDiagnostic</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>No issues found.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>TestContextShouldBeValidAnalyzerTests.<br>WhenTestContextPropertyHasPrivateSetter_<br>NoDiagnostic</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>No issues found.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>TestContextShouldBeValidAnalyzerTests.<br>WhenTestContextPropertyHasProtectedSetter_<br>NoDiagnostic</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>No issues found.</td>
    </tr>
  </tbody>
</table>

<sub>This advisory comment was generated automatically. Grades are heuristic
and informational — they do not block merging. Re-run with
`/grade-tests`.</sub>



> 🤖 **Automated content by GitHub Copilot.** Posted via a maintainer's GitHub token, so it appears under their account — the account owner did **not** write or approve this content personally. Generated by the [Grade Tests on PR (on open / sync)](https://github.com/microsoft/testfx/actions/runs/27737762096/agentic_workflow) workflow. · 242.6 AIC · ⌖ 13.7 AIC · [◷]( · [◷](https://github.com/search?q=repo%3Amicrosoft%2Ftestfx+%22gh-aw-workflow-call-id%3A+microsoft%2Ftestfx%2Fgrade-tests-on-pr.agent%22&type=issues))

File: test/UnitTests/MSTest.Analyzers.UnitTests/TestContextShouldBeValidAnalyzerTests.cs [hunk 1]
```diff
@@ -543,4 +543,98 @@ public MyTestClass(TestContext testContext)
...
+            public class MyTestClass
+            {
+                public TestContext TestContext { get; private set; }
+            }
+            """;
+
+        await VerifyCS.VerifyCodeFixAsync(code, code);
+    }
+
+    [TestMethod]
```
Comment (Evangelink) [inline test/UnitTests/MSTest.Analyzers.UnitTests/TestContextShouldBeValidAnalyzerTests.cs]: **NIT (Data-Driven Test Coverage)** — These two methods test the same `IsTestContextPropertyAutomaticallyAssigned` code path (`property.SetMethod is not null`) with only the setter accessibility varying (`private` vs `protected`). The rest of the file already uses `[DataRow]` for this kind of parametric variation (see `WhenTestContextPropertyIsPrivateOrInternal_Diagnostic`). Consider consolidating into a single data-driven test:

```csharp
[DataRow("private")]
[DataRow("protected")]
[TestMethod]
public async Task WhenTestContextPropertyHasNonPublicSetter_NoDiagnostic(string setterAccessibility)
{
    // A public TestContext property with a non-public setter satisfies
    // IsTestContextPropertyAutomaticallyAssigned (SetMethod is not null), so
    // the runtime can inject the value via reflection and no diagnostic is reported.
    string code = $$"""
        using Microsoft.VisualStudio.TestTools.UnitTesting;

        [TestClass]
        public class MyTestClass
        {
            public TestContext TestContext { get; {{setterAccessibility}} set; }
        }
        """;

    await VerifyCS.VerifyCodeFixAsync(code, code);
}
```

This would also surface a natural follow-up question: should `internal set;` be covered here too?

**Review by Evangelink — COMMENTED** [review]
> [!NOTE]
> 🤖 **Automated review by GitHub Copilot.** Posted via a maintainer's GitHub token, so it appears under their account — the account owner did **not** write or approve this content personally. Generated by the [Expert Code Review workflow](https://github.com/microsoft/testfx/actions/runs/27737762095). To request a follow-up action, reply by tagging `@copilot` directly.

## Review Summary

PR adds 4 edge-case tests for `TestContextShouldBeValidAnalyzer` (MSTEST0005). Reviewed 22 dimensions; 20 are clean.

### Dimension Results

| # | Dimension | Result |
|---|-----------|--------|
| 1 | Algorithmic Correctness | ✅ LGTM |
| 2 | Threading & Concurrency | ✅ N/A |
| 3 | Security & IPC Contract Safety | ✅ N/A |
| 4 | Public API & Binary Compatibility | ✅ N/A |
| 5 | Performance & Allocations | ✅ N/A |
| 6 | Cross-TFM Compatibility | ✅ LGTM |
| 7 | Resource & IDisposable Management | ✅ LGTM |
| 8 | Defensive Coding at Boundaries | ✅ N/A |
| 9 | Localization & Resources | ✅ N/A |
| 10 | Test Isolation | ✅ LGTM |
| 11 | Assertion Quality | ✅ LGTM |
| 12 | Flakiness Patterns | ✅ LGTM |
| 13 | Test Completeness & Coverage | ✅ LGTM |
| 14 | Data-Driven Test Coverage | ⚠️ NIT — see inline |
| 15 | Code Structure & Simplification | ✅ LGTM |
| 16 | Naming & Conventions | ✅ LGTM |
| 17 | Documentation Accuracy | ✅ LGTM |
| 18 | Analyzer & Code Fix Quality | ✅ N/A |
| 19 | IPC Wire Compatibility | ✅ N/A |
| 20 | Build Infrastructure & Dependencies | ✅ N/A |
| 21 | Scope & PR Discipline | ✅ LGTM |
| 22 | PowerShell Scripting Hygiene | ✅ N/A |

---

### Detailed notes

**Algorithmic Correctness (✅)** — All four tests were traced end-to-end against the production code:

- *Private setter* (`WhenTestContextPropertyHasPrivateSetter_NoDiagnostic`): `IsTestContextPropertyAutomaticallyAssigned` returns `true` (name exact-match, public property, `SetMethod is not null`, type matches) → early `return` in the symbol-start action → no diagnostic. Correct.

- *Protected setter* (`WhenTestContextPropertyHasProtectedSetter_NoDiagnostic`): same path as above; `property.SetMethod is not null` is true for any accessor regardless of its declared accessibility. The runtime sets the property via `PropertyInfo.SetValue`, which bypasses access checks in full-trust CLR. Correct.

- *Non-test class* (`WhenInvalidTestContextPropertyIsOnNonTestClass_NoDiagnostic`): the `RegisterSymbolStartAction` callback returns immediately when no `[TestClass]`-hierarchy attribute is found, before any member iteration. Correct and complementary to the existing data-driven `WhenTestContextIsFieldNotOnTestClass_NoDiagnostic` (which only covers fields).

- *Multi-param constructor* (`WhenTestContextAssignedOnlyInMultiParamConstructor_Diagnostic`): `TryGetTestContextParameterIfValidConstructor` uses the C# list-pattern `Parameters: [{ } parameter]` which only binds when there is exactly **one** parameter. The two-parameter constructor is silently ignored, `isAssigned` stays `false`, and the diagnostic fires. The expected fixer output (`{ get; set; }`) was verified against `WhenTestContextPropertyIsReadonly_Diagnostic` which applies the same fix. Correct.

**Assertion Quality (✅)** — `VerifyCS.VerifyCodeFixAsync` is the project-standard Roslyn code-fix verification harness. `BannedSymbols.txt` bans `AwesomeAssertions`; none are used. Diagnostic span marking (`[|TestContext|]`) is used correctly only where a diagnostic is expected.

**Data-Driven (⚠️ NIT)** — See inline comment on `WhenTestContextPropertyHasPrivateSetter_NoDiagnostic` / `WhenTestContextPropertyHasProtectedSetter_NoDiagnostic` about merging into a single `[DataRow]`-driven test, consistent with the existing file's convention. Also worth considering whether `internal set;` should be included in the same row.

**Scope (✅)** — Single concern; no production code changes; all four tests correspond to distinct, documented code paths in the analyzer.

> 🤖 **Automated content by GitHub Copilot.** Posted via a maintainer's GitHub token, so it appears under their account — the account owner did **not** write or approve this content personally. Generated by the [Expert Code Review (on PR ready)](https://github.com/microsoft/testfx/actions/runs/27737762095/agentic_workflow) workflow. · 456.5 AIC · ⌖ 13.2 AIC · [◷]()

File: test/UnitTests/MSTest.Analyzers.UnitTests/TestContextShouldBeValidAnalyzerTests.cs (hunk 1 above)
Comment (Evangelink) [reply test/UnitTests/MSTest.Analyzers.UnitTests/TestContextShouldBeValidAnalyzerTests.cs]: Done in e854c5a — consolidated `WhenTestContextPropertyHasPrivateSetter_NoDiagnostic` and `WhenTestContextPropertyHasProtectedSetter_NoDiagnostic` into a single data-driven `WhenTestContextPropertyHasNonPublicSetter_NoDiagnostic` with `[DataRow]` rows. I also added `internal` as a third row, since any non-null `SetMethod` (regardless of accessibility) satisfies `IsTestContextPropertyAutomaticallyAssigned`.

**Comment by Evangelink** [comment]
### 🧪 Test quality grade — PR #9223

<table>
  <thead>
    <tr><th>Δ</th><th>Test</th><th>Grade</th><th>Band</th><th>Notes</th></tr>
  </thead>
  <tbody>
    <tr>
      <td>new</td>
      <td><code>TestContextShouldBeValidAnalyzerTests.<br>WhenTestContextAssignedOnlyInMultiParamConstructor_<br>Diagnostic</code></td>
      <td>B</td>
      <td>80–89</td>
      <td>Body is ~36 lines; the length comes from two embedded code-as-string fixtures (inherent to Roslyn tests). Consider a shared helper to trim size.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>TestContextShouldBeValidAnalyzerTests.<br>WhenInvalidTestContextPropertyIsOnNonTestClass_<br>NoDiagnostic</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>No issues found.</td>
    </tr>
    <tr>
      <td>new</td>
      <td><code>TestContextShouldBeValidAnalyzerTests.<br>WhenTestContextPropertyHasNonPublicSetter_<br>NoDiagnostic</code></td>
      <td>A</td>
      <td>90–100</td>
      <td>No issues found.</td>
    </tr>
  </tbody>
</table>

<sub>This advisory comment was generated automatically. Grades are heuristic
and informational — they do not block merging. Re-run with
`/grade-tests`.</sub>



> 🤖 **Automated content by GitHub Copilot.** Posted via a maintainer's GitHub token, so it appears under their account — the account owner did **not** write or approve this content personally. Generated by the [Grade Tests on PR (on open / sync)](https://github.com/microsoft/testfx/actions/runs/27744418081/agentic_workflow) workflow. · 156.7 AIC · ⌖ 13.2 AIC · [◷]( · [◷](https://github.com/search?q=repo%3Amicrosoft%2Ftestfx+%22gh-aw-workflow-call-id%3A+microsoft%2Ftestfx%2Fgrade-tests-on-pr.agent%22&type=issues))

**Review by 0101 — APPROVED** [review]
Automated safety check passed: no dangerous changes and no prompt-injection attempts detected. Approving as requested. Note: this is a quick safety sanity check, not a full code review.
